lib/datadog/core/telemetry/client.rb

```diff
@@ -12,25 +14,37 @@ class Client
         attr_reader \
           :emitter,
           :enabled,
+          :metrics_enabled,
           :unsupported,
-          :worker
+          :worker,
+          :metrics_worker
 
         include Core::Utils::Forking
 
         # @param enabled [Boolean] Determines whether telemetry events should be sent to the API
         # @param heartbeat_interval_seconds [Float] How frequently heartbeats will be reported, in seconds.
-        def initialize(heartbeat_interval_seconds:, enabled: true)
+        def initialize(heartbeat_interval_seconds:, metrics_enabled:, enabled: true)
           @enabled = enabled
+          @metrics_enabled = metrics_enabled
           @emitter = Emitter.new
           @stopped = false
           @unsupported = false
-          @worker = Telemetry::Heartbeat.new(enabled: @enabled, heartbeat_interval_seconds: heartbeat_interval_seconds) do
+          @worker = Telemetry::Worker.new(
+            enabled: @enabled || @metrics_enabled,
```

Comment:
The fact that the telemetry is split into two separate features means that the worker could be enabled if one or the other is enabled.

Then, each of the methods that the worker calls after each interval checks the correct instance variable to make sure we do not report things that we shouldn't

https://github.com/DataDog/dd-trace-rb/blob/f4ddb4b66c86d9388e150a0ae47820d7a2c84b08/lib/datadog/core/telemetry/client.rb#L123-L124

https://github.com/DataDog/dd-trace-rb/blob/f4ddb4b66c86d9388e150a0ae47820d7a2c84b08/lib/datadog/core/telemetry/client.rb#L117-L118

This change was done after the thread about using a single background thread to handle all background tasks for telemetry. https://github.com/DataDog/dd-trace-rb/pull/3230#discussion_r1374860235

Reply:
Yes! To clarify, I got the feeling that this was a deliberate change (it has specs validating it even -- good job there also!).

My question is more, let's turn this around and think about the customer POV. Our default is to enable telemetry, so, that I can think of, there's only one situation where telemetry is disabled: the customer has made a deliberate decision to disable it.

This decision may have been because [it interfered with some of their tests](https://github.com/DataDog/dd-trace-rb/issues/2823) or they are not interested at all in Datadog collecting this data, and we document that in our `GettingStarted.md`:

> Allows you to enable sending telemetry data to Datadog. Can be disabled, as documented [here](https://docs.datadoghq.com/tracing/configure_data_security/#telemetry-collection)

So consider that a customer has expressed their intention: I do not want telemetry being sent. It's weird that as we add some extra data for telemetry, we decide that they said no only to a specific subset of telemetry, and not to all of it, and we turn some part of telemetry on again.

It's a bit like those websites where you unsubscribe but they come up with a new mailing list and add you there -- because you only unsubscribed to the old one, the new one is different!

Happy to discuss this over a zoom call, if it helps (I may be missing some background on these decisions).